x/superfluid/keeper/migrate.go

```diff
@@ -1,99 +1,302 @@
+	_, err = sdk.ValAddressFromBech32(valAddr)
+	if err != nil {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, err
+	}
+
+	// Superfluid undelegate the superfluid delegated position.
+	// This deletes the connection between the lock and the intermediate account, deletes the synthetic lock, and burns the synthetic osmo.
+	intermediateAccount, err := k.SuperfluidUndelegateToConcentratedPosition(ctx, sender.String(), lockId)
+	if err != nil {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, err
+	}
+
+	// Force unlock, validate the provided sharesToMigrate, and exit the balancer pool.
+	// This will return the coins that will be used to create the concentrated liquidity position.
+	exitCoins, err := k.validateSharesToMigrateUnlockAndExitBalancerPool(ctx, sender, poolIdLeaving, preMigrationLock, gammSharesInLock, sharesToMigrate)
+	if err != nil {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, err
+	}
+
+	// Create a full range (min to max tick) concentrated liquidity position, lock it, and superfluid delegate it.
+	positionId, amount0, amount1, liquidity, joinTime, concentratedLockId, err = k.clk.CreateFullRangePositionLocked(ctx, concentratedPool, sender, exitCoins, remainingLockTime)
+	if err != nil {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, err
+	}
+	err = k.SuperfluidDelegate(ctx, sender.String(), concentratedLockId, intermediateAccount.ValAddr)
+	if err != nil {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, err
+	}
+
+	// If there are remaining gamm shares, we must re-lock them.
+	remainingGammShares := gammSharesInLock.Sub(sharesToMigrate)
+	if !remainingGammShares.IsZero() {
+		// Create a new lock with the remaining gamm shares for the remaining lock time.
+		newLock, err := k.lk.CreateLock(ctx, sender, sdk.NewCoins(remainingGammShares), remainingLockTime)
+		gammLockId = newLock.ID
+		if err != nil {
+			return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, err
+		}
+
+		// If the gamm lock was previously superfluid bonded, superfluid delegate the gamm like normal
+		err = k.SuperfluidDelegate(ctx, sender.String(), gammLockId, valAddr)
+		if err != nil {
+			return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, err
+		}
+	}
+
+	return positionId, amount0, amount1, liquidity, joinTime, gammLockId, concentratedLockId, nil
+}
+
+// migrateSuperfluidUnbondingBalancerToConcentrated migrates a user's superfluid unbonding balancer position to a superfluid unbonding concentrated liquidity position.
+// The function force unlocks and exits the balancer pool, creates a full range concentrated liquidity position, and locks it. If there are any remaining gamm shares, they are re-locked and begin unlocking where they left off.
+// The function returns the concentrated liquidity position ID, amounts of tokens in the position, the liquidity amount, join time, and IDs of the involved pools and locks.
```

Comment:
Please also comment about the creation of the synth lock, why the old one becomes invalid and the new one is needed

Reply:
https://github.com/osmosis-labs/osmosis/pull/4885/commits/2c40ba0afadd23adead9e14fe1fb161d81f4c658